### PR TITLE
Fix issue related with forking processes

### PR DESCRIPTION
**Pull request by ansasaki**
When a process using the pkcs11 engine forks and the list of slots is changed in the system (e.g. by adding or removing devices), the slot ID's referenced by objects in the engine (e.g. keys) are not updated. When an operation is requested (e.g. signature), the old slot ID is used, leading to an error.

This adds a test to detect this behaviour. The test script follows these steps:

1. Create 2 devices using SoftHSM
2. Generate a key pair in the second SoftHSM and a certificate signed by the created private key
3. Run the test program in background that:
  a. Loads the engine and the generated private key
  b. Fork and make the child process wait for a signal (and the parent to wait for the child)
4. Delete the first created device, changing the list of slots
5. Send a signal to make the test program to continue
6. The test program will continue:
  a. The child process detects the fork, and reloads the context
  b. The context reload will not update the slot ID used by the key
  c. When the signature is called, it tries to use the old slot ID, causing an error

I believe this is related with #195 

The solution is to update the references to slot ID when a fork is detected.
Notice that the call to ``pkcs11_enumerate_slots()`` at p11_load.c, line 130, apparently does nothing to the context, since the obtained list is discarded.

### CODE REVIEW

**Comment by nmav** [comment]
It may be a good idea to enable travis for your repository so you can check whether your code would work prior to opening the PR on upstream.
https://travis-ci.org/ansasaki/libp11

File: tests/fork-change-slot.c [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+#include <signal.h>
+
+#include <execinfo.h>
+
+#include <openssl/evp.h>
+#include <openssl/x509.h>
+#include <openssl/bio.h>
+#include <openssl/pem.h>
+#include <openssl/engine.h>
+
```
Comment (nmav) [inline tests/fork-change-slot.c]: A small summary of what this program does would be nice. Think of it of a descritpion for someone trying to figure what this program does few years later.

File: tests/fork-change-slot.c [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+    if (cert == NULL) {
+        error_queue("PEM_read_bio_X509", pid);
+    }
+
+    BIO_free(bio);
+    return cert;
+}
+
+static void usage(char *argv[])
```
Comment (nmav) [inline tests/fork-change-slot.c]: no serious by why not only pass argv[0]?

File: tests/fork-change-slot.c [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+    engine = ENGINE_by_id("pkcs11");
+    if (engine == NULL) {
+        fprintf(stderr, "fatal: engine \"pkcs11\" not available\n");
+        error_queue("ENGINE_by_id", pid);
+        goto failed;
+    }
+
+    /* Initialize to get the engine functional reference */
+    if (ENGINE_init(engine)) {
+        //TODO Improve arguments parsing using getopt or argp
```
Comment (nmav) [inline tests/fork-change-slot.c]: I don't think it makes sense to have a todo in code which you send upstream. I would look like a todo for him. About the point itself, I'm not sure what would a full blown parser buy on a test suite tool.

File: tests/fork-change-slot.c [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+    }
+
+    /* Spawn processes and check child return */
+    if (spawn_processes(num_processes)) {
+        goto failed;
+    }
+    pid = getpid();
+
+    /* Get random bytes */
+    random = OPENSSL_malloc(RANDOM_SIZE);
```
Comment (nmav) [inline tests/fork-change-slot.c]: why not use a buffer in stack?

File: tests/fork-change-slot.c [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+    if (random == NULL)
+        goto failed;
+
+    fd = open(RANDOM_SOURCE, O_RDONLY);
```
Comment (nmav) [inline tests/fork-change-slot.c]: Why use /dev/urandom instead of RAND_bytes?

File: tests/fork-change-slot.c [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+                pid, rc, RANDOM_SIZE);
+        close(fd);
+        goto failed;
+    }
+
+    close(fd);
+
+    /* Ask for a hash of the random data, signed by the key */
+    siglen = MAX_SIGSIZE;
+    signature = OPENSSL_malloc(MAX_SIGSIZE);
```
Comment (nmav) [inline tests/fork-change-slot.c]: same question here about stack vs heap. It seems to me that code would be simpler if these move to stack.

File: tests/fork-change-slot.softhsm [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+outdir="output.$$"
+
+echo "Current directory: $(pwd)"
+echo "Source directory: ${srcdir}"
+echo "Output directory: ${outdir}"
+
+mkdir -p $outdir
+
+for i in /usr/lib64/pkcs11 /usr/lib/softhsm /usr/local/lib/softhsm /opt/local/lib/softhsm /usr/lib/x86_64-linux-gnu/softhsm /usr/lib /usr/lib64/softhsm;do
```
Comment (nmav) [inline tests/fork-change-slot.softhsm]: If these are re-used from another script would it make sense to move them in a common place and include it here?

File: tests/fork-change-slot.softhsm [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+# Initialize SoftHSM db
+init_db
+
+# Create 2 different tokens
+init_card $PIN $PUK token1
+init_card $PIN $PUK token2
+
+# Generate a key pair in the second token
+p11tool --generate-rsa --bits=2048 --label=pkey --login \
```
Comment (nmav) [inline tests/fork-change-slot.softhsm]: It may be that p11tool isn't available in travis, as it is not used by any other tests. You may want to switch to pkcs11-tool (which it is used by other tests) or make sure p11tool is installed in .travis.sh.

File: tests/fork-change-slot.softhsm [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+# Run the test program which will stop and wait for a signal (SIGUSR1)
+./fork-change-slot \
+    "pkcs11:token=token2;object=pkey;type=private;pin-value=$PIN" \
+    "$outdir//cert.crt" 2 &
+pid=$!
+
+# Wait the test program to reach the sigwait
+sleep 1
+
+# Remove the first token to change the list of slots
```
Comment (nmav) [inline tests/fork-change-slot.softhsm]: `to change the order of listed slots` may be more  clear

File: tests/fork-change-slot.softhsm [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    -keyform engine -out $outdir/cert.crt -x509 -subj "/CN=localhost"
+
+# Run the test program which will stop and wait for a signal (SIGUSR1)
+./fork-change-slot \
+    "pkcs11:token=token2;object=pkey;type=private;pin-value=$PIN" \
+    "$outdir//cert.crt" 2 &
+pid=$!
+
+# Wait the test program to reach the sigwait
+sleep 1
```
Comment (nmav) [inline tests/fork-change-slot.softhsm]: travis machine may be quite slow. I'd put a safer sleep of 3 or 5 seconds to be sure

**Review by nmav — CHANGES_REQUESTED** [review]
Seems ok, may need some clarifications and explicit testing with travis as indicated by the comments to be ready for submission.

File: tests/fork-change-slot.c (hunk 1 above)
Comment (ansasaki) [reply tests/fork-change-slot.c]: Ok. I'll add a description for the program

File: tests/fork-change-slot.c (hunk 2 above)
Comment (ansasaki) [reply tests/fork-change-slot.c]: Previously I was using more arguments to write the usage, but then I changed the program. Since I am not using the other elements of the list any more, I'll change this accordingly.

File: tests/fork-change-slot.c (hunk 3 above)
Comment (ansasaki) [reply tests/fork-change-slot.c]: About the point, the main reason was to use automatic generation of options like "-h" for help. But, since this code is for testing purposes, and it has a specific goal, it seems unnecessary. I'll remove all the TODO's in future codes that I am willing to submit to the upstream.

File: tests/fork-change-slot.c (hunk 4 above)
Comment (ansasaki) [reply tests/fork-change-slot.c]: I simply copied all the signing/verifying part from other test code (fork-test.c). But I can improve this here.

File: tests/fork-change-slot.c (hunk 5 above)
Comment (ansasaki) [reply tests/fork-change-slot.c]: Same reason as above.

File: tests/fork-change-slot.c (hunk 6 above)
Comment (ansasaki) [reply tests/fork-change-slot.c]: Again, same reason as above.

File: tests/fork-change-slot.softhsm (hunk 7 above)
Comment (ansasaki) [reply tests/fork-change-slot.softhsm]: Yes, I copied this part from other script, ``rsa-common.sh``, which is used in other tests. The problem is the way it is written in ``rsa-common.sh``, which is not so generic to be used here. I can improve the code there, to make more generic, but it would lead to changes in other tests as well. Would it be OK to change other non-related tests for this?

File: tests/fork-change-slot.softhsm (hunk 8 above)
Comment (ansasaki) [reply tests/fork-change-slot.softhsm]: I'll switch to pkcs11-tool to keep it consistent with the other tests.

File: tests/fork-change-slot.softhsm (hunk 10 above)
Comment (ansasaki) [reply tests/fork-change-slot.softhsm]: OK, I'll increase the sleep time to make sure it will work.

File: tests/fork-change-slot.softhsm (hunk 9 above)
Comment (ansasaki) [reply tests/fork-change-slot.softhsm]: I do not agree on this, since the goal is not to change the order of listed slots, but to change the slot to which the used device (``token2`` in my test) is associated. I would rephrase it to ``"Remove the fist token to change de slot ID associated with token2"``. Do you agree?

File: tests/fork-change-slot.softhsm (hunk 7 above)
Comment (nmav) [reply tests/fork-change-slot.softhsm]: I'd find that ok (and I hope that the maintainer would agree with me :)

File: tests/fork-change-slot.softhsm (hunk 9 above)
Comment (nmav) [reply tests/fork-change-slot.softhsm]: fine with me

**Comment by ansasaki** [comment]
Since the PR containing this reproducer and the fix for the issue was merged in upstream (#206) in , I'll close this.
